server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java

```diff
@@ -197,12 +267,121 @@ public void validate(final RelativeByteSizeValue value, final Map<Setting<?>, Ob
+                CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_FROZEN_MAX_HEADROOM_SETTING
             );
-            return settings.iterator();
+
+            // Ensure that if max headroom values are set, then watermark values are ratios/percentages.
+            final RelativeByteSizeValue low = (RelativeByteSizeValue) settings.get(CLUSTER_ROUTING_ALLOCATION_LOW_DISK_WATERMARK_SETTING);
+            if (low.isAbsolute()
+                && (lowHeadroom.equals(ByteSizeValue.MINUS_ONE) == false
+                    || highHeadroom.equals(ByteSizeValue.MINUS_ONE) == false
+                    || floodHeadroom.equals(ByteSizeValue.MINUS_ONE) == false)) {
+                // No need to check that the high or flood stage watermarks are absolute as well, since there is another check in
+                // WatermarkValidator that all low/high/flood watermarks should be either ratios/percentages or absolute values.
+                throw new IllegalArgumentException(
+                    "At least one of the disk max headroom settings is set [low="
+                        + lowHeadroom.getStringRep()
+                        + ", high="
+                        + highHeadroom.getStringRep()
+                        + ", flood="
+                        + floodHeadroom.getStringRep()
+                        + "], while the disk watermark values are set to absolute values instead of ratios/percentages, e.g., "
+                        + "the low watermark is ["
+                        + low.getStringRep()
+                        + "]"
+                );
+            }
+
+            // Similar check for the frozen flood watermark and max headroom settings
+            final RelativeByteSizeValue frozenFlood = (RelativeByteSizeValue) settings.get(
+                CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_FROZEN_WATERMARK_SETTING
+            );
+            if (frozenFlood.isAbsolute() && frozenFloodHeadroom.equals(ByteSizeValue.MINUS_ONE) == false) {
+                throw new IllegalArgumentException(
+                    "The frozen flood stage disk max headroom setting is set ["
+                        + frozenFloodHeadroom.getStringRep()
+                        + "], while the frozen flood stage disk watermark setting is set to an absolute value "
+                        + "instead of a ratio/percentage ["
+                        + frozenFlood.getStringRep()
+                        + "]"
+                );
+            }
+
+            if (lowHeadroom.equals(ByteSizeValue.MINUS_ONE) == false && highHeadroom.equals(ByteSizeValue.MINUS_ONE)) {
+                throw new IllegalArgumentException(
+                    "high disk max headroom ["
+                        + highHeadroom.getStringRep()
+                        + "] is not set, while the low disk max headroom is set ["
+                        + lowHeadroom.getStringRep()
+                        + "]"
+                );
+            }
```

Comment:
I think there is an issue with this validation. Imagine a user is upgrading a cluster from a version prior to this change to a version with it. 

The prior settings are:
```
cluster.routing.allocation.disk.watermark.low=null (is not set)
cluster.routing.allocation.disk.watermark.high=91% (any non-default value)
```

This way the new settings are initialized as:
```
cluster.routing.allocation.disk.watermark.low.max_headroom=200GB
cluster.routing.allocation.disk.watermark.high.max_headroom=-1
```
that would fail this validation.

I believe the similar could happen for below validation of highHeadroom & floodHeadroom.

Reply:
Hi @idegtiarenko, indeed, reading the code I believe you're right. This would make the user need to define the max headroom explicitly. Seems like an issue that could break upgrades indeed.

I think a solution out of this would be that the default values of the headrooms return -1 if any of the watermarks are explicitly set. Currently they return -1 only if the respective watermark is explicity set. So with this correction, it would mean that the low max headroom in your example would return -1 instead of 200GB.

Have you created a ticket for this?

Reply:
I believe this is covered by this issue: https://github.com/elastic/elasticsearch/issues/93468